[GFC][Integration] Begin allowing automatic placement with row auto flow.

sammygill opened the pull request:
#### f14c8d82e651277c1133a711ac7c5805f5c27ec3
<pre>
[GFC][Integration] Begin allowing automatic placement with row auto flow.
<a href="https://bugs.webkit.org/show_bug.cgi?id=303585">https://bugs.webkit.org/show_bug.cgi?id=303585</a>
<a href="https://rdar.apple.com/165061123">rdar://165061123</a>

Reviewed by Brandon Stewart.

This patch relaxes our restrictions for GFC with regards to item
placement to start allowing some form of auto placement to run. The
simplest styling to begin auto placement for is when the values for
grid-item-start and grid-item-end are both auto.

One tricky bit that we need to check alongside this, however, is if
there are potentially too many items locked to a given row. This is
because if there are too many items that are forced to a particular row
than there are number of columns in the explicit grid, then we will need
to potentially add columns to the implicit grid in order to accommodate
them. Since we are focusing on placement within the explicit grid for
now, we need to check for this case and bail if we find ourselves
potentially in that state.

* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
(WebCore::LayoutIntegration::hasValidColumnEnd):
New overload to help with the case where grid-column-start is auto. For
now, we only support grid-column-end: auto and bail in all other cases.

(WebCore::LayoutIntegration::gridLayoutAvoidanceReason):
To determine whether there are potentially too many items in a row, we
need to start keeping track of the number of grid items explicitly
placed in a row. We can do this by using a Vector where each index
corresponds to a row in the grid and the value is the number of items in
that row. While we are checking a grid item&apos;s row positioning, we will
increment the value that corresponds with the row it is in if it is
explicitly placed.

Canonical link: <a href="https://commits.webkit.org/305990@main">https://commits.webkit.org/305990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef0e1ab86d1d74f6988065e2c0367c72e0d3b203

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/139979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/12360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/1490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/148122 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/141852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/12512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/107175 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/142929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10067 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/125351 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88050 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/9718 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/7237 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/8411 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/118940 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/1354 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/150910 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/12158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/1420 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/115591 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/10318 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/115907 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29456 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/10714 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/121832 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/67050 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12087 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/1306 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/11828 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/75774 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12023 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/11875 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -122,6 +122,24 @@ static std::optional<GridAvoidanceReason> hasValidColumnEnd(const Style::GridPos
     );
 }
 
+static std::optional<GridAvoidanceReason> hasValidColumnEnd(const CSS::Keyword::Auto&,  const Style::GridPosition columnEnd)
```
I think this is effectively returning a bool

sammygill replied:
Changed to return a bool!

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -122,6 +122,24 @@ static std::optional<GridAvoidanceReason> hasValidColumnEnd(const Style::GridPos
+    );
+}
+
 static std::optional<GridAvoidanceReason> hasValidRowEnd(const Style::GridPositionExplicit&, const Style::GridPosition rowEnd, size_t linesFromGridTemplateRowsCount)
```
Here as well

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -338,8 +358,8 @@ static OptionSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid
-            [&](const CSS::Keyword::Auto&) -> std::optional<GridAvoidanceReason> {
-                return GridAvoidanceReason::GridItemHasUnsupportedColumnPlacement;
+            [&](const CSS::Keyword::Auto& autoPosition) -> std::optional<GridAvoidanceReason> {
+                return hasValidColumnEnd(autoPosition, gridItemStyle->gridItemColumnEnd());
```
auto& columnEnd = gridItemStyle->gridItemColumnEnd();

...


                return hasValidColumnEnd(autoPosition,columnEnd);

yulunwu2 commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
auto & rowEnd ...

yulunwu2 reviewed (approved):
a few nits but overall LGTM

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -384,6 +418,18 @@ static EnumSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid&
+        if (ineligibleRowIndex != notFound)
+            ADD_REASON_AND_RETURN_IF_NEEDED(GridItemHasUnsupportedRowPlacement, reasons, reasonCollectionMode);
+
+        ASSERT(renderGridStyle->gridAutoFlow().isRow(), "The above assumes row auto flow. If we start supporting column then it needs to be reworked.");
```
This is a grid level check; not item check. Can we do this outside the for loop?

sammygill replied:
Yup done

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -101,42 +101,60 @@ enum class GridAvoidanceReason : uint8_t {
 }
 
-static std::optional<GridAvoidanceReason> hasValidRowEnd(const Style::GridPositionExplicit&, const Style::GridPosition rowEnd, size_t linesFromGridTemplateRowsCount)
+static bool hasValidColumnEnd(const CSS::Keyword::Auto&,  const Style::GridPosition columnEnd)
```
Can we get rid of the first parameter? Also double space for second parameter between ",  const"

sammygill replied:
The idea behind the first parameter is to have different overloads for this function depending on what type of value we have for column start so that we can allow certain cases if we find implementing one scenario easier than the other. We could combine it into one generic function but the disadvantage of that would probably be nested switches which would decrease readability imo

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -101,42 +101,60 @@ enum class GridAvoidanceReason : uint8_t {
 
 
-static std::optional<GridAvoidanceReason> hasValidColumnEnd(const Style::GridPositionExplicit&, const Style::GridPosition columnEnd, size_t linesFromGridTemplateColumnsCount)
+static bool hasValidColumnEnd(const Style::GridPositionExplicit&, const Style::GridPosition columnEnd, size_t linesFromGridTemplateColumnsCount)
```
Same as below

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -367,9 +398,19 @@ static EnumSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid&
+                explicitlyPlacedItemsInRowCount.resize(rowStartLineNumber + 1);
+                ++explicitlyPlacedItemsInRowCount[rowStartLineNumber];
+                return { };
+
```
Extra blank line

sammygill replied:
Fixed

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -384,6 +425,17 @@ static EnumSet<GridAvoidanceReason> gridLayoutAvoidanceReason(const RenderGrid&
+        });
+        if (ineligibleRowIndex != notFound)
+            ADD_REASON_AND_RETURN_IF_NEEDED(GridItemHasUnsupportedRowPlacement, reasons, reasonCollectionMode);
+
```
Extra blank line

sammygill replied:
Fixed

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -101,42 +101,64 @@ enum class GridAvoidanceReason : uint8_t {
 }
 
-static std::optional<GridAvoidanceReason> hasValidRowEnd(const Style::GridPositionExplicit&, const Style::GridPosition rowEnd, size_t linesFromGridTemplateRowsCount)
+static bool hasValidColumnEnd(const CSS::Keyword::Auto& autoColumnStart, const Style::GridPosition columnEnd)
```
We can just remove the `autoColumnStart`, so we do not need the UNUSED_PARAM

sammygill replied:
I actually ended up adding these in the most recent iteration because I was looking back at the argument again and it wasn't immediately obvious to me what it represented by looking at the signature so I just added them as an additional annotation

stwrt commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
```diff
@@ -101,42 +101,64 @@ enum class GridAvoidanceReason : uint8_t {
 
 
-static std::optional<GridAvoidanceReason> hasValidColumnEnd(const Style::GridPositionExplicit&, const Style::GridPosition columnEnd, size_t linesFromGridTemplateColumnsCount)
+static bool hasValidColumnEnd(const Style::GridPositionExplicit& explicitColumnStart, const Style::GridPosition columnEnd, size_t linesFromGridTemplateColumnsCount)
```
We can just remove the `explicitColumnStart`, so we do not need the UNUSED_PARAM

sammygill committed:
[GFC][Integration] Begin allowing automatic placement with row auto flow.
https://bugs.webkit.org/show_bug.cgi?id=303585
rdar://165061123

Reviewed by Brandon Stewart.

This patch relaxes our restrictions for GFC with regards to item
placement to start allowing some form of auto placement to run. The
simplest styling to begin auto placement for is when the values for
grid-item-start and grid-item-end are both auto.

One tricky bit that we need to check alongside this, however, is if
there are potentially too many items locked to a given row. This is
because if there are too many items that are forced to a particular row
than there are number of columns in the explicit grid, then we will need
to potentially add columns to the implicit grid in order to accommodate
them. Since we are focusing on placement within the explicit grid for
now, we need to check for this case and bail if we find ourselves
potentially in that state.

* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
(WebCore::LayoutIntegration::hasValidColumnEnd):
New overload to help with the case where grid-column-start is auto. For
now, we only support grid-column-end: auto and bail in all other cases.

(WebCore::LayoutIntegration::gridLayoutAvoidanceReason):
To determine whether there are potentially too many items in a row, we
need to start keeping track of the number of grid items explicitly
placed in a row. We can do this by using a Vector where each index
corresponds to a row in the grid and the value is the number of items in
that row. While we are checking a grid item's row positioning, we will
increment the value that corresponds with the row it is in if it is
explicitly placed.

Canonical link: https://commits.webkit.org/305990@main
